daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -1125,67 +1125,50 @@ private[lf] final case class Compiler(
     * The returned free variables are de bruijn indices
     * adjusted to the stack of the caller. */
   def freeVars(expr: SExpr, initiallyBound: Int): Set[Int] = {
-    var bound = initiallyBound
-    var free = Set.empty[Int]
-
-    def go(expr: SExpr): Unit =
+    def go(expr: SExpr, bound: Int): Set[Int] =
```

Comment:
```suggestion
    def go(expr: SExpr, bound: Int, free: Set[Int]): Set[Int] =
```

Reply:
It looks like if you accept all @remyhaemmerle-da's changes, you can add `@tailrec` to this function too.

Reply:
:+1:

Reply:
It's a tree traversal so I don't think that will be possible. `args.foldLeft(free)((free, arg) =>  go(arg, bound, free))` for example is not a tail call.

Reply:
Missed that; thanks for explaining.